### DISCUSSION TITLE
RCP: How to set Radio transmit power as channel changes?

### DISCUSSION THREAD

**Discussion by rretanubun** [openthread/openthread#8766] [Q&A] [answered]
Hello,

My platform is a TME built on an RCP architecture on Thread stack 1.3.0

My RCP stack have support for `otPlatRadioGetTransmitPower()` and `otPlatRadioSetTransmitPower()`
and I can see the spinel messages being sent to the RCP.

`otPlatRadioSetChannelMaxTransmitPower()` is returning `NotImplemented`
because it maps to the weak function in `/src/code/radio/radio_platform.cpp`

As I understand it, the RCP vendor library also does not support pre-mapping Channel to TransmitPower
and expect to be told the transmit power to use for the channel in use.

In some regions of the world (e.g. North America) the allowed transmit power
is different per channel (for example `Ch-11-to-25 = 19 dBm and Ch-26 = 9.9 dBm`)
So it seems each time the device switches channels, it needs to check what channel it is switching to 
and call `otPlatRadioSetTransmitPower()` accordingly.

Outside of just `intentional operational dataset` changes, Thread have other mechanisms
that causes it to switch channels (e.g. `MLE Announcements`)

Is there a good place to hook in the call to `SetTransmitPower` to the RCP Radio
(e.g. https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel_impl.hpp#L2020)
to handle such cases (or is this handled already using a different mechanism?)

Thank you for everyone's time.

**Comment by jwhui** [comment, marked as answer]
Ideally, the radio vendor should provide drivers that implement the appropriate radio platform APIs, like `otPlatRadioSetChannelMaxTransmitPower`. This is what they are designed for. I would reach out to the radio vendor regarding implementation of these radio platform APIs.

**Reply by rretanubun** [reply]
Thanks Jonathan, I am following up on the vendor side as well. The piece that is not clear to me
 is the expected behavior that Thread needs the Radio platform APIs to perform.

For example, is the expectation `otPlatRadioSetChannelMaxTransmitPower` is called for channel 11-26 once at boot from the host-processor to the RCP and then as channels are used, the RCP will automatically apply the pre-configured TransmitPower?

Or is the `host processor` going to send `otPlatRadioSetChannelMaxTransmitPower` each time a channel is switched and the RCP just applies the latest settings?

I am not clear which entity (Host vs RCP) is expected to host the `channel to Transmit power` map during run time in this design.

**Reply by jwhui** [reply]
The radio driver should maintain this map. You can refer to the [nRF528xx implementation](https://github.com/openthread/ot-nrf528xx/blob/main/src/src/radio.c) as one example.
